### PR TITLE
5X: Use doubly-linked block lists in aset.c to reduce large-chunk overhead.

### PR DESCRIPTION
**Pull request by pivotal-ning-yu**
When loading data to a AOCO table with 300 columns and 1000 partitions we noticed that `pfree()` consumed most of the cpu, so we backported below patch from upstream:

> From ff97741bc810390db6dd4da0f31ee1e93c8d3abb Mon Sep 17 00:00:00 2001
> From: Tom Lane <tgl@sss.pgh.pa.us>
> Date: Wed, 8 Mar 2017 12:21:12 -0500
> Subject: [PATCH] Use doubly-linked block lists in aset.c to reduce large-chunk
>  overhead.
> 
> Large chunks (those too large for any palloc freelist) are managed as
> separate blocks.  Formerly, realloc'ing or pfree'ing such a chunk required
> O(N) time in a context with N blocks, since we had to traipse down the
> singly-linked block list to locate the block's predecessor before we could
> fix the list links.  This can result in O(N^2) runtime in situations where
> large numbers of such chunks are manipulated within one context.  Cases
> like that were not foreseen in the original design of aset.c, and indeed
> didn't arise until fairly recently.  But such problems can now occur in
> reorderbuffer.c and in hash joining, both of which make repeated large
> requests without scaling up their request size as they do so, and which
> will free their requests in not-necessarily-LIFO order.
> 
> To fix, change the block list from singly-linked to doubly-linked.
> This adds another 4 or 8 bytes to ALLOC_BLOCKHDRSZ, but that doesn't
> seem like unacceptable overhead, since aset.c's blocks are normally
> 8K or more, and never less than 1K in current practice.
> 
> In passing, get rid of some redundant AllocChunkGetPointer() calls in
> AllocSetRealloc (the compiler might be smart enough to optimize these
> away anyway, but no need to assume that) and improve AllocSetCheck's
> checking of block header fields.
> 
> Back-patch to 9.4 where reorderbuffer.c appeared.  We could take this
> further back, but currently there's no evidence that it would be useful.
> 
> Discussion: https://postgr.es/m/CAMkU=1x1hvue1XYrZoWk_omG0Ja5nBvTdvgrOeVkkeqs71CV8g@mail.gmail.com

----

We also included a fix for an `invalid alloc size` error during cache invalidation.
In AddInvalidationMessage() a new chunk always has double size of last
chunk, but once the size exceeds 1GB, the max allowed alloc size, an
error like "invalid memory alloc request size 1,342,177,300" will be
thrown. Fixed by limiting the chunk size.

----

FYI, here is the perf data for the `pfree()` slowness issue:

```
Samples: 18K of event 'cpu-clock:uhH', Event count (approx.): 4699250000
  Children      Self  Command   Shared Object       Symbol                                          ▒
-   99.84%     0.01%  postgres  postgres            [.] aocs_insert_finish                          ▒
   - 99.83% aocs_insert_finish                                                                      ◆
      - 99.68% close_ds_write (inlined)                                                             ▒
         - destroy_datumstreamwrite                                                                 ▒
            - 99.66% DatumStreamBlockWrite_Finish                                                   ▒
               - 99.63% AllocSetFree                                                                ▒
                    AllocSetFreeImpl (inlined)                                                      ▒
+   99.83%     0.00%  postgres  postgres            [.] CopyFrom                                    ▒
+   99.82%     0.00%  postgres  postgres            [.] DoCopyInternal                              ▒
+   99.80%     0.00%  postgres  postgres            [.] DoCopy                                      ▒
+   99.76%     0.00%  postgres  postgres            [.] PostgresMain                                ▒
+   99.76%     0.00%  postgres  postgres            [.] exec_mpp_query                              ▒
+   99.76%     0.00%  postgres  postgres            [.] PortalRun                                   ▒
+   99.76%     0.00%  postgres  postgres            [.] ProcessUtility                              ▒
+   99.76%     0.00%  postgres  postgres            [.] PortalRunMulti (inlined)                    ▒
+   99.76%     0.00%  postgres  postgres            [.] PortalRunUtility (inlined)                  ▒
+   99.75%     0.00%  postgres  postgres            [.] ServerLoop                                  ▒
+   99.75%     0.00%  postgres  postgres            [.] BackendStartup (inlined)                    ▒
+   99.75%     0.00%  postgres  postgres            [.] BackendRun (inlined)                        ▒
+   99.71%     0.00%  postgres  postgres            [.] _start                                      ▒
+   99.71%     0.00%  postgres  libc-2.27.so        [.] __libc_start_main                           ▒
+   99.71%     0.00%  postgres  postgres            [.] main                                        ▒
+   99.71%     0.00%  postgres  postgres            [.] PostmasterMain                              ▒
+   99.68%     0.00%  postgres  postgres            [.] close_ds_write (inlined)                    ▒
+   99.68%     0.00%  postgres  postgres            [.] destroy_datumstreamwrite                    ▒
+   99.66%     0.01%  postgres  postgres            [.] DatumStreamBlockWrite_Finish                ▒
+   99.63%    99.60%  postgres  postgres            [.] AllocSetFree                                ▒
+   99.63%     0.00%  postgres  postgres            [.] AllocSetFreeImpl (inlined)                  ▒
     0.29%     0.00%  postgres  [unknown]           [.] 0xffffffffffffffff                          ▒
     0.13%     0.01%  postgres  postgres            [.] MirroredAppendOnly_FlushAndClose            ▒
     0.13%     0.00%  postgres  postgres            [.] AppendOnlyStorageWrite_TransactionFlushAndCl▒
     0.12%     0.00%  postgres  postgres            [.] datumstreamwrite_close_file                 ▒
     0.11%     0.00%  postgres  postgres            [.] AppendOnlyStorageWrite_FlushAndCloseFile (in▒
     0.09%     0.01%  postgres  postgres            [.] FileRepPrimary_IsOperationCompleted         ▒
     0.08%     0.01%  postgres  postgres            [.] FileRepAckPrimary_IsOperationCompleted      ▒
     0.07%     0.00%  postgres  postgres            [.] BufferedAppendCompleteFile                  ▒
     0.07%     0.00%  postgres  postgres            [.] BufferedAppendWrite (inlined)               ▒
     0.07%     0.00%  postgres  postgres            [.] MirroredAppendOnly_Append                   ▒
     0.05%     0.00%  postgres  libc-2.27.so        [.] _IO_vfprintf_internal (inlined)             ▒
     0.05%     0.00%  postgres  postgres            [.] FileRepPrimary_MirrorFlushAndClose          ▒
     0.05%     0.00%  postgres  libc-2.27.so        [.] __GI___snprintf (inlined)                   ▒
     0.04%     0.02%  postgres  postgres            [.] FileWrite                                   ▒
     0.04%     0.01%  postgres  postgres            [.] FileRepPrimary_ConstructAndInsertMessage    ▒
```

## Background
The patches were first merged on 4X for a hotfix release.  Now we are porting them to 5X.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/utils/mmgr/aset.c [hunk 1]
```diff
@@ -1362,7 +1362,7 @@ AllocSetFreeImpl(MemoryContext context, void *pointer, bool isHeader)
 		 * just past the chunk.
 		 */
 		if (block->aset != set ||
-			block->freeptr != block->endptr ||
+			block->freeptr != UserPtr_GetEndPtr(block) ||
```
Comment (asimrp) [inline src/backend/utils/mmgr/aset.c]: These changes belong to a different commit.  Can you please check if they are still needed and if so, create a separate commit for them?  I couldn't see them in the original upstream commit: ff97741bc810390db6dd4da0f31ee1e93c8d3abb

File: src/backend/utils/cache/inval.c [hunk 2]
```diff
@@ -233,6 +233,14 @@ AddInvalidationMessage(InvalidationChunk **listHdr,
 		/* Need another chunk; double size of last chunk */
 		int			chunksize = 2 * chunk->maxitems;
 
+		/*
+		 * Keep in mind: the max allowed alloc size is about 1GB, for
+		 * simplification we set the upper limit to half of that.
+		 */
+#define MAXCHUNKSIZE (MaxAllocSize / 2 / sizeof(SharedInvalidationMessage))
```
Comment (berlin-ab) [inline src/backend/utils/cache/inval.c]: This looks similar to https://github.com/greenplum-db/gpdb/pull/9113

What's the relationship?

File: src/backend/utils/mmgr/aset.c (hunk 1 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/mmgr/aset.c]: In postgres the `endptr` was introduced in https://github.com/postgres/postgres/commit/ead64f317be6eae7cdff9074659f8140aea3c4d5, which was included since postgres 6.5; in gpdb `endptr` was replaced with `UserPtr_GetEndPtr()` since 42ca3506ee3b00d054dc31c9cb0d59b69e70e51d, so I have to do the same replacement while back porting the patch.

File: src/backend/utils/cache/inval.c (hunk 2 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/cache/inval.c]: Both #9113 and the current one is forward-porting of a patchset from gpdb 4X branch.  We had to do this because we need the patchset in 4X for a hotfix release, there was no time to run the normal "master->4X" back porting process at that time.

So as the patchset is in 4X now, we are now forward-porting them to master, 6X and 5X.  In master and 6X the upstream patch is already there during the postgres merging process, so #9113 only contains one patch while the current one, which is for 5X, contains two.

File: src/backend/utils/mmgr/aset.c (hunk 1 above)
Comment (pivotal-ning-yu) [reply src/backend/utils/mmgr/aset.c]: > The cherry-picked from commit hash listed in "Fix invalid alloc size when invalidating caches" commit message is from 4.3_STABLE (proprietary) branch. Can you please mention it explicitly? Otherwise, it confuses an open source developer when they run `git show` against it.

Makes sense, I'll update the commit message

Comment (pivotal-ning-yu) [reply src/backend/utils/mmgr/aset.c]: Rebased with current 5X branch and updated the commit message.
